src/binder_data.c

```diff
@@ -0,0 +1,2452 @@
+    call->addresses = binder_strv_from_hidl_string_vec(&dc->addresses);
+    call->pcscf = binder_strv_from_hidl_string_vec(&dc->pcscf);
+
+    DBG("[status=%d,retry=%d,cid=%d,active=%d,type=%d,ifname=%s,"
+        "mtu=%d,address=%s,dns=%s,gateways=%s,pcscf=%s]",
+        call->status, call->retry_time, call->cid, call->active,
+        dc->type, call->ifname, call->mtu,
+        binder_print_strv(call->addresses, " "),
+        binder_print_strv(call->dnses, " "),
+        binder_print_strv(call->gateways, " "),
+        binder_print_strv(call->pcscf, " "));
+    return call;
+}
+
+static
+GSList*
+binder_data_call_list_1_4(
+    const RadioDataCall_1_4* calls,
+    gsize n)
+{
+    if (n) {
+        gsize i;
+        GSList* l = NULL;
+
+        DBG("num=%u", (guint) n);
+        for (i = 0; i < n; i++) {
+            l = g_slist_insert_sorted(l, binder_data_call_new_1_4(calls + i),
+                binder_data_call_compare);
+        }
+        return l;
+    } else {
+        DBG("no data calls");
+        return NULL;
+    }
+}
+
+static
+gboolean
+binder_data_call_equal(
+    const BinderDataCall* c1,
+    const BinderDataCall* c2)
+{
+    if (!c1 && !c2) {
+        return TRUE;
+    } else if (c1 && c2) {
+        return c1->cid == c2->cid &&
+            c1->status == c2->status &&
+            c1->active == c2->active &&
+            c1->prot == c2->prot &&
+            c1->retry_time == c2->retry_time &&
+            c1->mtu == c2->mtu &&
+            !g_strcmp0(c1->ifname, c2->ifname) &&
+            gutil_strv_equal(c1->dnses, c2->dnses) &&
+            gutil_strv_equal(c1->gateways, c2->gateways) &&
+            gutil_strv_equal(c1->addresses, c2->addresses) &&
+            gutil_strv_equal(c1->pcscf, c2->pcscf);
+    } else {
+        return FALSE;
+    }
+}
+
+static
+gboolean
+binder_data_call_list_equal(
+    GSList* l1,
+    GSList* l2)
+{
+    while (l1 && l2) {
+        if (!binder_data_call_equal(l1->data, l2->data)) {
+            return FALSE;
+        }
+        l1 = l1->next;
+        l2 = l2->next;
+    }
+    return (!l1 && !l2);
+}
+
+static
+gboolean
+binder_data_call_list_contains(
+    GSList* l,
+    const BinderDataCall* call)
+{
+    while (l) {
+        if (binder_data_call_equal(l->data, call)) {
+            return TRUE;
+        }
+        l = l->next;
+    }
+    return FALSE;
+}
+
+/* extern */
+BinderDataCall*
+binder_data_call_find(
+    GSList* l,
+    int cid)
+{
+    while (l) {
+        BinderDataCall* call = l->data;
+
+        if (call->cid == cid) {
+            return call;
+        }
+        l = l->next;
+    }
+    return NULL;
+}
+
+static
+void
+binder_data_set_calls(
+    BinderDataObject* self,
+    GSList* list)
+{
+    BinderBase* base = &self->base;
+    BinderData* data = &self->pub;
+    GHashTableIter it;
+    gpointer key;
+
+    if (binder_data_call_list_equal(data->calls, list)) {
+        binder_data_call_list_free(list);
```

Comment:
It can, if the `grab` table has changed since the last time we've been here. The number of data calls is never large, an extra consistency check wouldn't be much of an overhead, I would rather leave it as is. And again, this part was copied from the old ofono-ril-plugin which has gone through a lot of testing and real life use - when in doubt I usually leave such code unmodified.